mesonbuild/mintro.py

```diff
@@ -298,48 +304,103 @@ def list_projinfo_from_source(sourcedir):
-        list_buildoptions(coredata, builddata)
-    elif options.tests:
-        list_tests(testdata)
-    elif options.benchmarks:
-        list_tests(benchmarkdata)
-    elif options.dependencies:
-        list_deps(coredata)
-    elif options.projectinfo:
-        list_projinfo(builddata)
-    else:
+    # Load build data to make sure that the version matches
+    # TODO Find a better solution for this
+    cdata.load(options.builddir)
+
+    results = []
+    toextract = []
+
+    if options.all or options.benchmarks:
+        toextract += ['benchmarks']
+    if options.all or options.buildoptions:
+        toextract += ['buildoptions']
+    if options.all or options.buildsystem_files:
+        toextract += ['buildsystem_files']
+    if options.all or options.dependencies:
+        toextract += ['dependencies']
+    if options.all or options.list_installed:
+        toextract += ['installed']
+    if options.all or options.projectinfo:
+        toextract += ['projectinfo']
+    if options.all or options.list_targets:
+        toextract += ['targets']
+    if options.target_files is not None:
+        targets_file = os.path.join(infodir, 'intro-targets.json')
+        with open(targets_file, 'r') as fp:
+            targets = json.load(fp)
+        builddata = build.load(options.builddir) # TODO remove this in a breaking changes PR
+        results += [list_target_files(options.target_files, targets, builddata)]
+    if options.all or options.tests:
+        toextract += ['tests']
+
+    for i in toextract:
+        curr = os.path.join(infodir, 'intro-{}.json'.format(i))
+        if not os.path.isfile(curr):
+            print('Introspection file {} does not exist.'.format(curr))
+            return 1
+        with open(curr, 'r') as fp:
+            results += [(i, json.load(fp))]
+
+    indent = options.indent if options.indent > 0 else None
+
+    if len(results) == 0 and not options.force_dict:
         print('No command specified')
         return 1
+    elif len(results) == 1 and not options.force_dict:
+        # Make to keep the existing output format for a single option
+        print(json.dumps(results[0][1], indent=indent))
+    else:
+        out = {}
+        for i in results:
+            out[i[0]] = i[1]
+        print(json.dumps(out, indent=indent))
     return 0
+
+def write_intro_info(intro_info, info_dir):
+    for i in intro_info:
+        out_file = os.path.join(info_dir, 'intro-{}.json'.format(i[0]))
+        with open(out_file, 'w') as fp:
+            json.dump(i[1], fp)
+
+def generate_introspection_file(builddata: build.Build, backend: backends.Backend):
+    coredata = builddata.environment.get_coredata()
+    benchmarkdata = backend.create_test_serialisation(builddata.get_benchmarks())
+    testdata = backend.create_test_serialisation(builddata.get_tests())
+    installdata = backend.create_install_data()
+
+    intro_info = [
+        list_benchmarks(benchmarkdata),
+        list_buildoptions(coredata),
```

Comment:
Can we move buildoptions to the top and add a comment that this should be first.
That way we know if buildoptions is older (or same date) as the any of the other ones that meson did reconfigure the build directory. (opposed to a meson configure which only changes buildoptions)

Reply:
I think it would be good to have something simple here as well. Not every consumer of this data is an IDE. But if someone wanted to check from a shell script if the options have changed it would be trivial with time stamps. (i.e. test intro-buildoptions -nt intro-projectinfo.json).

But i think it would be nice to have a info.json with something more suitable for IDE consumption too.

Reply:
This use case could also be solved with a `grep '"build_up_to_data": true' meson-info.json`. This should also be less error-prone.

Reply:
grepping in json is a really fragile way to do things. In that case timestamps are definitively more robust.

Reply:
Unless we make the JSON humanly readable (`json.dump(indent=2)`) and/or we also generate a `meson-info.sh` that can be sourced.

Reply:
This wouldn't really be a shell script. More like a config file that can also coincidentally be sourced by bash :)

This would then be more robust than a `test f1 -nt f2` and can contain even more useful information for the end user.

Something like https://gist.github.com/mensinda/c617025b3c5fbc00ff0455579c097ca0

Reply:
No shell sourcing. Ever!

Also, everything we provide must work identically on platforms that do not have a unix shell.

Reply:
OK, no bash script/config file generation.

A user can also use [jq](https://stedolan.github.io/jq/) to parse the generated JSON files in bash.

Reply:
We mostly agreed on irc that we can thing about this minor aspect after this PR. As it stands, there is not easy way to know if after `meson configure` the other introspection files have been regenerated or not. Which is exactly the situation we currently have. (you could look at timestamps of coredata.dat and the ninja build file, but that would be cheating)